# HG changeset patch
# User Wormnest
# Date 1377338985 -7200
#      Sat Aug 24 12:09:45 2013 +0200
# Node ID 458a6431c6c730396fd009116aed9313df403ae4
# Parent  c6f5e728a61e4ad2344175ba00dcf6603cbcf554
Add: Ctrl+Click in news history list scrolls to location of news.
Refactors go to location code from message window OnClick into a separate
function and then also uses that function in message history OnClick.
Changes the message history tooltip to specify the Ctrl+Click function.

diff -r c6f5e728a61e -r 458a6431c6c7 src/lang/english.txt
--- a/src/lang/english.txt	Wed Aug 21 17:45:24 2013 +0000
+++ b/src/lang/english.txt	Sat Aug 24 12:09:45 2013 +0200
@@ -744,7 +744,7 @@
 
 # News message history
 STR_MESSAGE_HISTORY                                             :{WHITE}Message History
-STR_MESSAGE_HISTORY_TOOLTIP                                     :{BLACK}A list of the recent news messages
+STR_MESSAGE_HISTORY_TOOLTIP                                     :{BLACK}A list of the recent news messages. Click on a message to show it. Ctrl+Click to go to the location of the message.
 STR_MESSAGE_NEWS_FORMAT                                         :{STRING}  -  {STRING5}
 
 STR_NEWS_MESSAGE_CAPTION                                        :{WHITE}Message
diff -r c6f5e728a61e -r 458a6431c6c7 src/news_gui.cpp
--- a/src/news_gui.cpp	Wed Aug 21 17:45:24 2013 +0000
+++ b/src/news_gui.cpp	Sat Aug 24 12:09:45 2013 +0200
@@ -74,6 +74,31 @@
 	}
 }
 
+/**
+ * Scroll to location of news article.
+ * @param ni The news item.
+ * @param use_extra_viewport When true open a new viewport unless it's a vehicle, 
+ * else scroll main window to location.
+ */
+void ScrollToLocation(const NewsItem *ni, bool use_extra_viewport = false)
+{
+	if (ni->reftype1 == NR_VEHICLE) {
+		const Vehicle *v = Vehicle::Get(ni->ref1);
+		ScrollMainWindowTo(v->x_pos, v->y_pos, v->z_pos);
+	} else {
+		TileIndex tile1 = GetReferenceTile(ni->reftype1, ni->ref1);
+		TileIndex tile2 = GetReferenceTile(ni->reftype2, ni->ref2);
+		if (use_extra_viewport) {
+			if (tile1 != INVALID_TILE) ShowExtraViewPortWindow(tile1);
+			if (tile2 != INVALID_TILE) ShowExtraViewPortWindow(tile2);
+		} else {
+			if ((tile1 == INVALID_TILE || !ScrollMainWindowToTile(tile1)) && tile2 != INVALID_TILE) {
+				ScrollMainWindowToTile(tile2);
+			}
+		}
+	}
+}
+
 /* Normal news items. */
 static const NWidgetPart _nested_normal_news_widgets[] = {
 	NWidget(WWT_PANEL, COLOUR_WHITE, WID_N_PANEL),
@@ -430,21 +455,7 @@
 				break; // Ignore clicks
 
 			default:
-				if (this->ni->reftype1 == NR_VEHICLE) {
-					const Vehicle *v = Vehicle::Get(this->ni->ref1);
-					ScrollMainWindowTo(v->x_pos, v->y_pos, v->z_pos);
-				} else {
-					TileIndex tile1 = GetReferenceTile(this->ni->reftype1, this->ni->ref1);
-					TileIndex tile2 = GetReferenceTile(this->ni->reftype2, this->ni->ref2);
-					if (_ctrl_pressed) {
-						if (tile1 != INVALID_TILE) ShowExtraViewPortWindow(tile1);
-						if (tile2 != INVALID_TILE) ShowExtraViewPortWindow(tile2);
-					} else {
-						if ((tile1 == INVALID_TILE || !ScrollMainWindowToTile(tile1)) && tile2 != INVALID_TILE) {
-							ScrollMainWindowToTile(tile2);
-						}
-					}
-				}
+				ScrollToLocation(this->ni, _ctrl_pressed);
 				break;
 		}
 	}
@@ -1078,7 +1089,13 @@
 				if (ni == NULL) return;
 			}
 
-			ShowNewsMessage(ni);
+			if (_ctrl_pressed) {
+				/* When ctrl is pressed don't show the news message but go to it's location. */
+				ScrollToLocation(ni, false);
+			}
+			else {
+				ShowNewsMessage(ni);
+			}
 		}
 	}
 
